### PR TITLE
Some changes for codacy

### LINKED ISSUE #277: Failing BasisSet tests
**Describe the bug**
The unit tests for `TestBasisSet::test_functions_by_shell` and `TestBasisSet::test_primitives_by_shell` are failing. After some digging it seems that it might be from a difference in the data type of the index (`int` and `'categorical'`).

I noticed that there is a `self.groupby` in the `BasisSet.functions_by_shell` and `BasisSet.primitives_by_shell` methods. What I'm thinking is that the `'categorical'` data type persists through the `groupby` and is causing this issue.

I propose that we add the line `obj = self._revert_categories(inplace=False)` in both methods and change the code in the `DataFrame._revert_categories` to

``` python
def _revert_categories(self, inplace=True):
    """
    Inplace conversion to categories.
    """
    if inplace:
        for column, dtype in self._categories.items():
            if column in self.columns:
                self[column] = self[column].astype(dtype)
    else:
        copy = self.copy()
        for column, dtype in copy._categories.items():
            if column in copy.columns:
                copy[column] = copy[column].astype(dtype)
        return copy
```

Essentially adding a kwarg so the replacement is not done inplace. I think this is a better way so that we don't have to run `_revert categories` and `_set_categories` for such a simple thing.

### PR DESCRIPTION
**Pull request by herbertludowieg**
These are just changes for codacy to stop complaining and to test the github actions.

Will go through all of the issues flagged by Codacy and try to fix what I can.

Version locked ipywidgets so tests can pass, see #276.

Changed behavior of `_revert_categories` so `BasisSet` tests can pass, closes #277.

### CODE REVIEW

File: exatomic/core/gradient.py [hunk 1]
```diff
@@ -23,10 +23,10 @@ def stats(self):
             norms = np.linalg.norm(data[cols].values, axis=1)
             rms = np.sqrt(np.mean(np.square(norms)))
             avg = np.mean(norms)
-            max = norms.max()
-            min = norms.min()
```
Comment (avmarchenko) [inline exatomic/core/gradient.py:27]: Thank you for these fixes.

File: .github/workflows/branch.yml [hunk 2]
```diff
@@ -62,12 +60,18 @@ jobs:
...
-        uses: actions/download-artifact@v2
-        with:
-          name: code-coverage-report-${{ needs.coverage-artifact-name.outputs.ref }}
+      - uses: actions/checkout@v2
       - name: Run codacy-coverage-reporter
         uses: codacy/codacy-coverage-reporter-action@v1
         with:
           project-token: ${{ secrets.CODACY_PROJECT_TOKEN }}
           coverage-reports: coverage.xml
+      #- name: Download Coverage Report
```
Comment (avmarchenko) [inline .github/workflows/branch.yml:69]: Remove commented out config?

File: exatomic/adf/tape21.py [hunk 3]
```diff
@@ -220,6 +213,7 @@ def parse_frequency(self, cart=True):
         df['r_mass'] = np.repeat(r_mass, nat)
         df['symbol'] = symbol
         df['label'] = label
+        # TODO: find out if this is stored in the file anywhere
```
Comment (avmarchenko) [inline exatomic/adf/tape21.py:216]: Create issue?

**Review by avmarchenko — APPROVED** [review]
Suggested non-blocking actions. Thank you for this work.

File: .github/workflows/branch.yml (hunk 2 above)
Comment (herbertludowieg) [reply .github/workflows/branch.yml:69]: Honestly, I have been trying some different things to get codacy coverage to trigger and report correctly.

Currently, it's having trouble verifying the `CODACY_PROJECT_KEY`.

Comment (avmarchenko) [reply .github/workflows/branch.yml:69]: Messaged privately to check the value you configured.

File: exatomic/adf/output.py [hunk 4]
```diff
@@ -707,21 +707,20 @@ def parse_electric_dipole(self):
...
         tdm['excitation'] -= 1
         diff = np.setdiff1d(df['excitation'].values.flatten(),
                             tdm['excitation'].values.flatten())
         if len(diff) > 0:
-            for idx in range(len(diff)):
-                df1 = tdm.loc[range(diff[idx])]
-                df2 = tdm.loc[range(diff[idx], tdm.shape[0])]
-                new_line = [diff[idx], df.loc[diff[idx], 'energy_ev'],
-                            df.loc[diff[idx], 'osc']]+[0.0]*int(len(tdm.columns)-3)
+            for idx, d in enumerate(diff):
```
Comment (avmarchenko) [inline exatomic/adf/output.py:719]:
```suggestion
            for d in diff:
```

**Comment by herbertludowieg** [comment]
Ready to be merged.

**Pull request merged**